### DISCUSSION TITLE
Terraform Provider

### DISCUSSION THREAD

**Discussion by psteinroe** [supabase/supabase#5166] [Feature Requests] [answered]
## Update: [work started here](https://github.com/supabase/terraform-provider-supabase)

Hey,

I have been working with Supabase for a few weeks now and have not been disappointed once - great work!

Although there is some stuff that has to be fixed beforehand, such as environment variables and the Postgres role concept revamp (e.g. https://github.com/supabase/cli/issues/120), a Terraform Provider for Supabase would be a great addition to a close-to-perfect continuous deployment setup. 

This would include the management of projects and their settings. 

I know this is probably quite a few month out, but wanted to leave it here to maybe discuss and gather use-cases.

Best
Philipp

**Comment by rorymalcolm** [comment]
Would be absolutely huge if this was offered IMO, really the only thing that scares me off using the platform

**Comment by dnnsthnnr** [comment]
There is already a first effort done here. Currently only support local environments: https://github.com/supabase/supabase/pull/5820

**Reply by jeremyj563** [reply]
#5820 is about deploying a local (development) instance of Supabase using some Terraform scripts. Whereas the OP here is talking about a _Terraform Provider_ built to configure and manage the state of a Supabase instance natively through it's API.

A Provider like this would make it possible to create and store all configuration for Supabase in a Git repository. It's a great way to track configuration changes made across the team as well as control drift and define configuration policies, etc..

We do this for large platforms like [GitLab](https://registry.terraform.io/providers/gitlabhq/gitlab/latest/docs) and [KeyCloak](https://registry.terraform.io/providers/mrparkers/keycloak/latest/docs), so if such a provider existed for Supabase it would be a very natural choice for us to use it too.

**Comment by liamkinne** [comment]
Hey all,

I'd be really interested in working on this and contributing it back to the project. The main blocker I can see is that the [management API](https://api.supabase.com/api/v1) is not permissive enough to allow a Terraform provider to be truly effective. It would be awesome to see some of the core endpoints (organizations, projects, etc) support `PATCH` and `DELETE` to allow the full range of CRUD operations that Terraform expects.

**Reply by liamkinne** [reply]
I have started working on this since it looks like the management API now makes this feasible (thanks SB team). Once it reaches an MVP usable state, I'll post again here. I'm also open to this being moved to the supabase-community org.

https://github.com/liamkinne/terraform-provider-supabase

**Comment by sweatybridge** [comment, marked as answer]
Hello, I'm collecting feedback for the alpha release of our official [terraform provider](https://github.com/supabase/terraform-provider-supabase) for hosted projects. For a start, it supports managing the following resources
- `supabase_project`
- `supabase_branch`
- `supabase_settings` (only `api` and `auth` at the moment)

A complete [CI/CD example](https://github.com/supabase/supabase-action-example) using GitHub Action is also included.

As with all alpha software, some things may not work as expected. I've documented some of the common workflows and caveats in [the tutorial](https://github.com/supabase/terraform-provider-supabase/blob/main/docs/tutorial.md). If you find any bugs or missing features, please feel free to leave a comment or open a GitHub issue.

**Reply by cdata** [reply]
I would really like to be able to derive the connection-pooled database URI from the `supabase_project` resource. Do you think it would be possible to do that?

**Reply by sweatybridge** [reply]
Yup, that's a good idea. I will work on pooler config this week.

**Reply by sweatybridge** [reply]
Pooler URL is available in provider 1.2.0, e.g. https://github.com/supabase/supabase-action-example/blob/main/supabase/remotes/production.tf#L17

**Reply by savelichalex** [reply]
@sweatybridge sorry, but what is `supabase_branch` and `supabase_settings` for?

**Reply by sweatybridge** [reply]
`supabase_settings` is used to configure both your project settings and branch settings in code. Instead of editing from the [settings page](https://supabase.com/dashboard/project/_/settings/database) on dashboard, you may find it easier to commit them in code and let terraform sync the settings up to your hosted project.

`supabase_branch` is used to create remote [development branches](https://supabase.com/docs/guides/platform/branching) without installing the Supabase GitHub integration. A common use case is if you want to run everything in your github action instead of relying on Supabase hosted runners.

**Reply by BarnabyShearer** [reply]
I would love to be able to pay you more! Seriously, just had an instance where I forgot to manually upgrade the compute after adding a new database via TF.

**Reply by ghost** [reply]
Would be cool to have a bit more detailed docs about how the supabase_settings resource works, I wanted to set up network restrictions but can't figure out how to define it in the resource

**Reply by sweatybridge** [reply]
Thanks for the feedback! I will prioritise implementing `compute` and `network restrictions` settings.

**Reply by sweatybridge** [reply]
Added support for these in provider v1.4.0
- Set compute instance size when creating [supabase_project](https://registry.terraform.io/providers/supabase/supabase/latest/docs/resources/project) resource.
- Add network restrictions to [supabase_settings](https://registry.terraform.io/providers/supabase/supabase/latest/docs/resources/settings) resource.

**Comment by maxcan** [comment]
Overall love it.  True step towards full, mature IaC support.

1. need support for the new JWT Signing keys.  Our backend logic is in AWS Lambda.  We can use the symmetric service_role_key export for now but thats deprecated and need a path forward with the signing keys.
2. the settings fields need more documentation. ideally they'd be broken out into terraform types, but at least a typescript or zod type showing what those objects should look like would help.
